### PR TITLE
Python3

### PR DESCRIPTION
**Pull request by umeier**
Test ran sucessfull on local machine with python 2.7 and 3.4.

### CODE REVIEW

**Comment by loicgasser** [comment]
Great I ll have a look at it this week !

File: quantized_mesh_tile/bbsphere.py [hunk 1]
```diff
@@ -25,7 +33,7 @@ def fromPoints(self, points):
             raise Exception('Your list of points must contain at least 2 points')
 
         nbPositions = len(points)
-        for i in xrange(0, nbPositions):
+        for i in range(0, nbPositions):
```
Comment (loicgasser) [inline quantized_mesh_tile/bbsphere.py]: I'd prefer to keep a generator in python 2.7, maybe:
`from past.builtins import xrange`
this way we make sure we always use generators

File: quantized_mesh_tile/bbsphere.py [hunk 2]
```diff
@@ -80,7 +88,7 @@ def fromPoints(self, points):
         naiveCenter = c3d.multiplyByScalar(c3d.add(minBoxPt, maxBoxPt), 0.5)
         naiveRadius = 0.0
 
-        for i in xrange(0, nbPositions):
+        for i in range(0, nbPositions):
```
Comment (loicgasser) [inline quantized_mesh_tile/bbsphere.py]: `from past.builtins import xrange`

File: quantized_mesh_tile/terrain.py [hunk 3]
```diff
@@ -329,17 +339,17 @@ def fromStringIO(self, f, hasLighting=False, hasWatermask=False):
         hd = 0
         # Vertices
         vertexCount = unpackEntry(f, TerrainTile.vertexData['vertexCount'])
-        for i in xrange(0, vertexCount):
+        for i in range(0, vertexCount):
```
Comment (loicgasser) [inline quantized_mesh_tile/terrain.py]: could we keep the lazy behaviour in python 2 as well?

File: quantized_mesh_tile/terrain.py [hunk 4]
```diff
@@ -320,7 +330,7 @@ def fromStringIO(self, f, hasLighting=False, hasWatermask=False):
         self.hasLighting = hasLighting
         self.hasWatermask = hasWatermask
         # Header
-        for k, v in TerrainTile.quantizedMeshHeader.iteritems():
+        for k, v in TerrainTile.quantizedMeshHeader.items():
```
Comment (loicgasser) [inline quantized_mesh_tile/terrain.py]: here it's not a problem if not lazy as we're iterating over a small dict.

File: quantized_mesh_tile/terrain.py [hunk 5]
```diff
@@ -329,17 +339,17 @@ def fromStringIO(self, f, hasLighting=False, hasWatermask=False):
...
         # Vertices
         vertexCount = unpackEntry(f, TerrainTile.vertexData['vertexCount'])
-        for i in xrange(0, vertexCount):
+        for i in range(0, vertexCount):
             ud += zigZagDecode(
                 unpackEntry(f, TerrainTile.vertexData['uVertexCount'])
             )
             self.u.append(ud)
-        for i in xrange(0, vertexCount):
+        for i in range(0, vertexCount):
```
Comment (loicgasser) [inline quantized_mesh_tile/terrain.py]: keep lazy

File: quantized_mesh_tile/terrain.py [hunk 6]
```diff
@@ -329,17 +339,17 @@ def fromStringIO(self, f, hasLighting=False, hasWatermask=False):
...
             vd += zigZagDecode(
                 unpackEntry(f, TerrainTile.vertexData['vVertexCount'])
             )
             self.v.append(vd)
-        for i in xrange(0, vertexCount):
+        for i in range(0, vertexCount):
```
Comment (loicgasser) [inline quantized_mesh_tile/terrain.py]: lazy

File: quantized_mesh_tile/terrain.py [hunk 7]
```diff
@@ -488,19 +498,19 @@ def _writeTo(self, f):
         f.write(
             packEntry(TerrainTile.vertexData['uVertexCount'], zigZagEncode(self.u[0]))
         )
-        for i in xrange(0, vertexCount - 1):
+        for i in range(0, vertexCount - 1):
```
Comment (loicgasser) [inline quantized_mesh_tile/terrain.py]: lazy

File: quantized_mesh_tile/terrain.py [hunk 8]
```diff
@@ -488,19 +498,19 @@ def _writeTo(self, f):
...
             ud = self.u[i + 1] - self.u[i]
             f.write(packEntry(TerrainTile.vertexData['uVertexCount'], zigZagEncode(ud)))
         f.write(
             packEntry(TerrainTile.vertexData['uVertexCount'], zigZagEncode(self.v[0]))
         )
-        for i in xrange(0, vertexCount - 1):
+        for i in range(0, vertexCount - 1):
```
Comment (loicgasser) [inline quantized_mesh_tile/terrain.py]: lazy

File: quantized_mesh_tile/terrain.py [hunk 9]
```diff
@@ -488,19 +498,19 @@ def _writeTo(self, f):
...
             vd = self.v[i + 1] - self.v[i]
             f.write(packEntry(TerrainTile.vertexData['vVertexCount'], zigZagEncode(vd)))
         f.write(
             packEntry(TerrainTile.vertexData['uVertexCount'], zigZagEncode(self.h[0]))
         )
-        for i in xrange(0, vertexCount - 1):
+        for i in range(0, vertexCount - 1):
```
Comment (loicgasser) [inline quantized_mesh_tile/terrain.py]: lazy

File: quantized_mesh_tile/terrain.py [hunk 10]
```diff
@@ -287,27 +297,27 @@ def _computeVerticesCoordinates(self):
...
                     lerp(
                         self.header['minimumHeight'],
                         self.header['maximumHeight'],
-                        float(h) / MAX
+                        old_div(float(h), MAX)
                     )
                 )
 
-    def fromStringIO(self, f, hasLighting=False, hasWatermask=False):
+    def from_BytesIO(self, f, hasLighting=False, hasWatermask=False):
```
Comment (loicgasser) [inline quantized_mesh_tile/terrain.py]: I'd prefer: `fromBytesIO`

File: quantized_mesh_tile/terrain.py [hunk 11]
```diff
@@ -430,9 +440,9 @@ def fromFile(self, filePath, hasLighting=False, hasWatermask=False, gzipped=Fals
         with open(filePath, 'rb') as f:
             if gzipped:
                 f = ungzipFileObject(f)
-            self.fromStringIO(f, hasLighting=hasLighting, hasWatermask=hasWatermask)
+            self.from_BytesIO(f, hasLighting=hasLighting, hasWatermask=hasWatermask)
 
-    def toStringIO(self, gzipped=False):
+    def to_BytesIO(self, gzipped=False):
```
Comment (loicgasser) [inline quantized_mesh_tile/terrain.py]: I'd prefer `toBytesIO`

File: quantized_mesh_tile/terrain.py [hunk 12]
```diff
@@ -549,7 +559,7 @@ def _writeTo(self, f):
             f.write(packEntry('B', 1))
 
             metaV = TerrainTile.OctEncodedVertexNormals
-            for i in xrange(0, vertexCount - 1):
+            for i in range(0, vertexCount - 1):
```
Comment (loicgasser) [inline quantized_mesh_tile/terrain.py]: lazy

File: quantized_mesh_tile/terrain.py [hunk 13]
```diff
@@ -569,7 +579,7 @@ def _writeTo(self, f):
                         'Unexpected number of rows for the watermask: %s' % nbRows
                     )
                 # From North to South
-                for i in xrange(0, nbRows):
+                for i in range(0, nbRows):
```
Comment (loicgasser) [inline quantized_mesh_tile/terrain.py]: lazy

File: quantized_mesh_tile/topology.py [hunk 14]
```diff
@@ -161,7 +164,7 @@ def _loadGeometry(self, geometrySpec):
         """
         A private method to convert a (E)WKB or (E)WKT to a Shapely geometry.
         """
-        if geometrySpec.startswith('POLYGON Z'):
+        if type(geometrySpec) is str and geometrySpec.startswith('POLYGON Z'):
```
Comment (loicgasser) [inline quantized_mesh_tile/topology.py]: why is that necessary `type(geometrySpec)`?

File: quantized_mesh_tile/utils.py [hunk 15]
```diff
@@ -149,7 +155,7 @@ def computeNormals(vertices, faces):
     areasPerFace = [0.0] * numFaces
     normalsPerVertex = np.zeros(vertices.shape, dtype=vertices.dtype)
 
-    for i in xrange(0, numFaces):
+    for i in range(0, numFaces):
```
Comment (loicgasser) [inline quantized_mesh_tile/utils.py]: keep lazy behaviour in py2 as well

File: quantized_mesh_tile/utils.py [hunk 16]
```diff
@@ -175,29 +181,29 @@ def computeNormals(vertices, faces):
         else:
             normalsPerFace[i] = normalB
 
-    for i in xrange(0, numFaces):
+    for i in range(0, numFaces):
```
Comment (loicgasser) [inline quantized_mesh_tile/utils.py]: lazy

File: quantized_mesh_tile/utils.py [hunk 17]
```diff
@@ -175,29 +181,29 @@ def computeNormals(vertices, faces):
...
         face = faces[i]
         weightedNormal = [c * areasPerFace[i] for c in normalsPerFace[i]]
         for j in face:
             normalsPerVertex[j] = c3d.add(normalsPerVertex[j], weightedNormal)
 
-    for i in xrange(0, numVertices):
+    for i in range(0, numVertices):
```
Comment (loicgasser) [inline quantized_mesh_tile/utils.py]: lazy

File: quantized_mesh_tile/utils.py [hunk 18]
```diff
@@ -209,7 +215,7 @@ def getCoordsIndex(n, i):
 # Creates all the potential pairs of coords
 def createCoordsPairs(l):
     coordsPairs = []
-    for i in xrange(0, len(l)):
+    for i in range(0, len(l)):
```
Comment (loicgasser) [inline quantized_mesh_tile/utils.py]: lazy

**Comment by loicgasser** [comment]
This is very good, thank you. I have only minor comments.
I would just like to keep the lazy behaviour in py2 as well.
I think this deserves a new version and will gladly deploy it when these comments are addressed.

File: quantized_mesh_tile/topology.py (hunk 14 above)
Comment (umeier) [inline quantized_mesh_tile/topology.py]: If type of geometrySpec is bytes we will have an exception here. So first check type then content.

**Comment by umeier** [comment]
I´m out of office next week, will address your comments asap.

File: quantized_mesh_tile/topology.py (hunk 14 above)
Comment (loicgasser) [inline quantized_mesh_tile/topology.py]: then `geometrySpec.startswith('POLYGON Z')`  should be
`geometrySpec.startswith(b'POLYGON Z')`
This should work both on py3 and py2.
Here you're simply ignoring it when in py3.

File: tests/test_terrain_tile.py [hunk 19]
```diff
@@ -252,10 +255,10 @@ def testGzippedTileCreationFromTopology(self):
         self.assertEqual(tile._east, 1.0)
         self.assertEqual(tile._north, 1.0)
 
-        fileLike = tile.toStringIO(gzipped=True)
-        self.assertIsInstance(fileLike, cStringIO.OutputType)
+        fileLike = tile.toBytesIO(gzipped=True)
+        self.assertIsInstance(fileLike, io.BytesIO)
 
-    def testFromStringIO(self):
+    def test_fromBytesIO(self):
```
Comment (loicgasser) [inline tests/test_terrain_tile.py]: I'll adapt that.

File: quantized_mesh_tile/topology.py (hunk 14 above)
Comment (loicgasser) [inline quantized_mesh_tile/topology.py]: ignore that.
